chitchat/src/listener.rs

```diff
@@ -0,0 +1,246 @@
+use std::collections::{BTreeMap, HashMap};
+use std::ops::Bound;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, RwLock, Weak};
+
+pub struct ListenerHandle {
+    prefix: String,
+    idx: usize,
+    listeners: Weak<RwLock<InnerListeners>>,
+}
+
+impl ListenerHandle {
+    // By default, a listener is cancelled when its handle is dropped.
+    // Calling forever prevents that.
+    //
+    // The listener itself will only be dropped when the Listeners object is dropped.
+    pub fn forever(mut self) {
+        self.listeners = Weak::new();
+    }
+}
+
+impl Drop for ListenerHandle {
+    fn drop(&mut self) {
+        if let Some(listeners) = self.listeners.upgrade() {
+            let mut listeners_guard = listeners.write().unwrap();
+            listeners_guard.remove_listener(&self.prefix, self.idx);
+        }
+    }
+}
+
+type BoxedListener = Box<dyn Fn(&str, &str) + 'static + Send + Sync>;
+
+#[derive(Default, Clone)]
+pub(crate) struct Listeners {
+    inner: Arc<RwLock<InnerListeners>>,
+}
+
+impl Listeners {
+    #[must_use]
+    pub(crate) fn subscribe_event(
+        &self,
+        key_prefix: impl ToString,
+        callback: impl Fn(&str, &str) + 'static + Send + Sync,
+    ) -> ListenerHandle {
+        let key_prefix = key_prefix.to_string();
+        let boxed_listener = Box::new(callback);
+        self.subscribe_event_for_ligher_monomorphization(key_prefix, boxed_listener)
+    }
+
+    fn subscribe_event_for_ligher_monomorphization(
+        &self,
+        key_prefix: String,
+        boxed_listener: BoxedListener,
+    ) -> ListenerHandle {
+        let key_prefix = key_prefix.to_string();
+        let weak_listeners = Arc::downgrade(&self.inner);
+        let mut inner_listener_guard = self.inner.write().unwrap();
+        let new_idx = inner_listener_guard
+            .listener_idx
+            .fetch_add(1, Ordering::SeqCst);
```

Comment:
If you're never reading the atomic variable, `Ordering::Relaxed` will do just fine.

Reply:
yes for an id we can do with relaxed.